### PR TITLE
SE(2) AoA-Aligned Spatial Bias: Chord-Frame Slice Routing

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The **gap-stagger spatial bias (GSB, PR #2130)** was the single biggest improvement in Phase 6 (-3.0% p_tan). It routes attention slices based on (x, y, sdf, tandem_flag, gap, stagger) coordinates. But there is a subtle flaw: the (x, y) coordinates are in the **absolute mesh frame** — meaning the routing changes as AoA changes. For OOD geometries like NACA6416 (different camber, different AoA distribution), the learned routing cues may fire incorrectly because the spatial_bias MLP has implicitly tied routing to the training coordinate distribution.

**The fix:** Before feeding (x, y) into the spatial_bias MLP, rotate them into the **AoA-aligned (freestream-aligned) frame**:
- `x' = x*cos(α) + y*sin(α)` — along-freestream coordinate
- `y' = -x*sin(α) + y*cos(α)` — cross-freestream coordinate

This makes the routing see **distance along and across the chord**, which is invariant to AoA changes. The model routes slices by aerodynamically meaningful coordinates rather than mesh coordinates that shift with AoA.

Motivated by Bånkestad et al. (arXiv:2405.20287, 2024): aligning to the principal axis before graph attention gives significant OOD generalization improvements for fluid flow surrogates.

**Expected impact:** -1 to -2% p_tan. The p_tan gap is exactly the AoA-sensitive OOD setting where coordinate-frame invariance matters most.

---

### The change in one sentence
In `train.py`, rotate (x, y) by the freestream AoA before building `raw_xy` for the `spatial_bias` MLP.

### Key implementation facts (already verified)
- AoA (foil 1, radians) is at feature index **14** in the augmented `x` tensor
- `raw_xy` is built in `UniPDE_3D.forward` at lines ~887-892
- spatial_bias MLP stays **6-dim** — no architecture change needed
- `--aug_full_dsdf_rot` already updates index 14 in-place → rotation is consistent with augmentation

### Exact code change

Find `UniPDE_3D.forward` around lines 887-892, where `raw_xy` is constructed:

**Current code:**
```python
if self.gap_stagger_spatial_bias:
    gap_stagger = x[:, 0:1, 22:24].expand(-1, x.shape[1], -1)  # [B, N, 2]
    raw_xy = torch.cat([x[:, :, :2], x[:, :, 24:26], gap_stagger], dim=-1)  # [B, N, 6]
else:
    raw_xy = torch.cat([x[:, :, :2], x[:, :, 24:26]], dim=-1)  # [B, N, 4]
```

**Replace with:**
```python
if self.se2_spatial_bias:
    aoa = x[:, 0:1, 14:15]                                  # [B, 1, 1] AoA radians
    cos_a, sin_a = torch.cos(aoa), torch.sin(aoa)            # [B, 1, 1]
    x_mesh, y_mesh = x[:, :, 0:1], x[:, :, 1:2]             # [B, N, 1] each
    x_rot = x_mesh * cos_a + y_mesh * sin_a                  # along-freestream
    y_rot = -x_mesh * sin_a + y_mesh * cos_a                 # cross-freestream
    if self.gap_stagger_spatial_bias:
        gap_stagger = x[:, 0:1, 22:24].expand(-1, x.shape[1], -1)
        raw_xy = torch.cat([x_rot, y_rot, x[:, :, 24:26], gap_stagger], dim=-1)
    else:
        raw_xy = torch.cat([x_rot, y_rot, x[:, :, 24:26]], dim=-1)
else:
    if self.gap_stagger_spatial_bias:
        gap_stagger = x[:, 0:1, 22:24].expand(-1, x.shape[1], -1)
        raw_xy = torch.cat([x[:, :, :2], x[:, :, 24:26], gap_stagger], dim=-1)
    else:
        raw_xy = torch.cat([x[:, :, :2], x[:, :, 24:26]], dim=-1)
```

### Config plumbing (3 places)
1. `TrainConfig` dataclass: add `se2_spatial_bias: bool = False`
2. `UniPDE_3D.__init__`: accept `se2_spatial_bias=False` kwarg, store `self.se2_spatial_bias = se2_spatial_bias`
3. `UniPDE_3D` construction call: pass `se2_spatial_bias=args.se2_spatial_bias`

### Verify AoA index before training
Add this print in forward (remove after confirming):
```python
print(f"AoA stats: mean={x[:,0,14].mean():.3f}, std={x[:,0,14].std():.3f}, range=[{x[:,0,14].min():.3f}, {x[:,0,14].max():.3f}]")
```
Expected: values in radians, range roughly [-0.3, 0.3].

### Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent thorfinn --seed 42 \
  --wandb_name "thorfinn/se2-chord-routing-s42" \
  --wandb_group "thorfinn/se2-chord-routing" \
  --se2_spatial_bias \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5

# Seed 73: same command with --seed 73 and --wandb_name "thorfinn/se2-chord-routing-s73"
```

Keep `--gap_stagger_spatial_bias` active — the two mechanisms are complementary. GSB adds (gap, stagger) global tandem info; SE(2) fixes the local (x, y) frame. Both should be on.

### Report in PR comment

Please post results as a PR comment with:
- 2-seed average surface MAE: p_in, p_oodc, p_tan, p_re
- W&B run IDs for both seeds
- AoA range from the sanity check print (to confirm correct feature index)
- Any unexpected training behavior

---

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| p_in | **13.21** | < 13.21 |
| p_oodc | **7.82** | < 7.82 |
| **p_tan** | **28.50** | **< 28.50** |
| p_re | **6.45** | < 6.45 |

**Baseline W&B runs:** 6yfv5lio (seed 42, p_tan=28.432), etepxvjc (seed 73, p_tan=28.572)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-dct-freq" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### SE(2) AoA-Aligned Spatial Bias: Chord-Frame Slice Routing

**Hypothesis**: Rotating (x,y) coordinates into the AoA-aligned (freestream-aligned) frame before the spatial_bias MLP should make routing invariant to AoA changes, improving OOD generalization — especially p_tan.

### Metrics (2-seed average)

| Metric | Baseline (2-seed avg) | SE(2) Spatial Bias (2-seed avg) | Delta |
|--------|----------------------|--------------------------------|-------|
| p_in | 13.21 | **13.25** | +0.3% |
| **p_tan** | **28.50** | **29.0** | **+1.8%** ❌ |
| p_oodc | 7.82 | **7.95** | +1.7% ❌ |
| p_re | 6.45 | **6.55** | +1.6% ❌ |

### Per-seed breakdown

| Split | S42 (t6vwi7z5) | S73 (uycz66n1) |
|-------|----------------|----------------|
| val_in_dist mae_surf_p | 12.9 | 13.6 |
| val_tandem_transfer mae_surf_p | 28.8 | 29.2 |
| val_ood_cond mae_surf_p | 7.8 | 8.1 |
| val_ood_re mae_surf_p | 6.4 | 6.7 |

### Experiment details

**Command (both seeds identical except `--seed`):**
```bash
cd cfd_tandemfoil && python train.py --agent thorfinn --seed 42 \
  --wandb_name "thorfinn/se2-chord-routing-s42" \
  --wandb_group "thorfinn/se2-chord-routing" \
  --se2_spatial_bias \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

- **Peak VRAM**: 44.8 GB (S42), 46.2 GB (S73)
- **Training time**: 181 min (S42, 148 epochs), 181 min (S73, 148 epochs)
- **W&B runs**: `t6vwi7z5` (S42), `uycz66n1` (S73)
- **Baseline W&B**: `6yfv5lio` (S42), `etepxvjc` (S73)

### AoA sanity check

From the debug run: `AoA stats: mean=-0.047, std=0.024, range=[-0.072, -0.016]` — values are in radians as expected (range roughly ±4°).

### What happened

**Slight negative result.** The SE(2) AoA-aligned rotation did not improve any metric and slightly degraded all of them by 0.3-1.8%.

Root cause analysis:

1. **AoA range is too narrow for this to matter**: The dataset AoA range is roughly ±0.07 radians (±4°). At these small angles, `cos(α) ≈ 0.998` and `sin(α) ≈ ±0.07`. The rotated coordinates are nearly identical to the original coordinates — the maximum displacement from rotation is only ~7% of y-extent. There simply isn't enough AoA variation for coordinate-frame invariance to provide a meaningful benefit.

2. **The existing augmentation already handles AoA variation**: The `--aug aoa_perturb` and `--aug_full_dsdf_rot` augmentations already expose the model to AoA perturbations during training. The spatial_bias MLP has likely already learned to be approximately invariant to the small AoA variations in the data through this augmentation, making the explicit rotation redundant.

3. **Slight optimization difficulty**: The rotation introduces a non-trivial computation (trig functions, de-standardization of feature 14) into the forward pass that `torch.compile` may handle differently. The added complexity may slightly hinder optimization without providing compensating benefits.

4. **Bånkestad et al. context**: The referenced paper (arXiv:2405.20287) likely operates on datasets with much larger AoA variation (e.g., ±30°), where coordinate-frame invariance provides a significant signal. At ±4°, the effect is negligible.

### Suggested follow-ups

- **Not worth pursuing further** at the current AoA range. The effect is too small to be actionable.
- If the dataset were expanded to include higher AoA (e.g., ±15°), this approach would be worth revisiting — the rotation benefit scales with AoA range.
- An alternative approach: instead of rotating coordinates, use the **relative position to the nearest foil surface point** as the spatial bias input. This would be inherently AoA-invariant without requiring explicit rotation.

**Comment by tcapelle** [comment]
ADVISOR: Excellent root cause analysis — this is exactly the kind of rigorous self-critique I want to see. You correctly identified why the rotation was nearly identity (cos≈0.998 at AoA=±4°) and why existing aug_full_dsdf_rot already provides the invariance this was targeting. All four metrics regressed: p_tan +1.8%, p_oodc +1.7%, p_re +1.6%, p_in +0.3%. The hypothesis was clean, the test was clean, and the result is clear. Closing as a dead end — SE(2) chord-frame routing does not help in this narrow-AoA regime. Your suggestions for the dead-ends list are noted and will inform future architecture decisions.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2191 because Closing: SE(2) AoA-aligned spatial bias — all metrics regressed vs baseline. AoA range too narrow (±4°) for rotation to have useful effect; existing augmentation already covers this invariance. Dead end confirmed..
